### ISSUE TITLE
Problems with Phobius

### ISSUE THREAD

**Issue by simas232** [nextgenusfs/funannotate#82]
I remember that Phobius was working properly a few months earlier. However, Phobius now doesn't work as it should. When I just simply try to run funannotate annotate, I get the following error in the end of the log file:

```
<TRUNCATED>
[12:06:35 AM]: Predicting secreted and transmembrane proteins using Phobius
Traceback (most recent call last):
File "<TRUNCATED>/funannotate/util/phobius-multiproc.py", line 47, in <module>
lib.setupLogging(log_name)
File "<TRUNCATED>/funannotate/lib/library.py", line 463, in setupLogging
fhnd = logging.FileHandler(LOGNAME)
File "<TRUNCATED>/miniconda2/envs/py27/lib/python2.7/logging/__init__.py", line 913, in __init__
StreamHandler.__init__(self, self._open())
File "<TRUNCATED>/miniconda2/envs/py27/lib/python2.7/logging/__init__.py", line 943, in _open
stream = open(self.baseFilename, self.mode)
IOError: [Errno 2] No such file or directory: '<TRUNCATED>/logfiles/phobius.log'
```

Regarding the files in the working directory, annotation_misc didn't contain any files related to Phobius. The directory 'logfiles' was also missing.

Then I tried to use the local installation of phobius.pl. I tested it as standalone function and it worked fine with my proteome. So after adding it to PATH I re-ran funannotate annotate but still got exactly the same error and the same file structure in the working directory.

Another way how I tried to fix this issue was to create 'logfiles' directory in the working directory. As a result, funannotate-functional.log now contained only "Predicting secreted and transmembrane proteins using Phobius" row without any trailing errors. Unfortunately, Phobius annotation was never finished, so I terminated it after 3 hours of waiting. I checked the files in the working directory again. The log file phobius.log was generated in logfiles folder, containing the single row:

> "2017-05-29 21:20:26,773: <TRUNCATED>/funannotate/util/phobius-multiproc.py -i <TRUNCATED>/annotate_misc/genome.proteins.fasta -o annotate_misc/phobius.results.txt -e simmarc@chalmers.se -l logfiles/phobius.log". 

I also found the new directory, called phobius_2409, which contained only the separate fasta files for each query protein. There were no differences in annotate_misc folder comparing to the previous runs.

My last idea was that perhaps phobius.pl file didn't have sufficient permissions, so I enabled read, write and execution permissions to phobius.pl and decodeanhmm, (chmod a+rwx FILE), but it didn't help as well, since I got the same output as before. Changing the first line of phobius.pl into "#!/usr/bin/env perl" didn't help either.

These whole results were consistent on two different machines, so I think it must be somehow related to the parts where funannotate tries to detect local Phobius installation and/or feeding the proteins to remote Phobius search. I checked the status of remote Phobius search server and it seems to work well. Thank you for the help in advance!

**Comment by nextgenusfs** [comment]
which version of funanntoate?  There was a bug where logfiles folder wasn't created correctly, as well there was a bug in parsing the phobius results in some instances it would exit prematurely. https://github.com/nextgenusfs/funannotate/commit/34858828653214aeaeed652549e67b4fe808a879.  This just seems like error in logfiles not created correctly, at least that is the python error. What was the funannotate command?

**Comment by simas232** [comment]
I am using funannotate v0.6.3

**Comment by nextgenusfs** [comment]
try passing the parent directory of 'predict_results', i.e. your output option from funannotate predict.  I'll try to have a look at why this happening, i'm traveling this whole week so will have limited time and spotty internet connectivity.

**Comment by nextgenusfs** [comment]
I might have just seen similar behavior.  I forgot to pass the `--email` parameter and since I did not have local phobius installed, I didn't get an error but rather appeared like it was working, when in fact it was not as the remote search requires email address. 

Did changing the input to funannotate annotate fix the problem?

**Comment by simas232** [comment]
I tried to do a clean run with _funannotate predict_ and _funannotate annotate_ and got the following error:

`[12:09:42 PM]: Predicting secreted and transmembrane proteins using Phobius
Traceback (most recent call last):
  File "/home/linuxbrew/funannotate/util/phobius-multiproc.py", line 84, in <module>
    result = line[1].split(' ')
IndexError: list index out of range`

It seems that the error is exactly the same like MichaelFokinNZ got two days ago. The pipeline doesn't stop, it still does all the downstreat processes, starting with InterProScan. Needless to say, I am using the latest commit of funannotate. 

I will try to re-run the pipeline without email setting and with InterProScan xml file

**Comment by nextgenusfs** [comment]
Okay.  I can fix this so it runs to completion, but would be better to determine what is causing it to choke.  Can you tell if phobius has completed successfully?  Is the file that it choked on empty?  Did anything get written to annotate_misc/phobius.results.txt?

**Comment by simas232** [comment]
Yes, the contents are slightly different than before. For example, the file with phobius intermediate results (_phobius_{random_number}_ directory) now contains not only protein fasta file, but also e.g. KKMAI_000001.phobius file. Unfortunately, all such *.phobius files are empty, containing only "SEQENCE ID                     TM SP PREDICTION". Regarding phobius.results.txt file in {organism}_output/annotate_misc directory, it contains only the line "ID	TM	SP	Prediction".

**Comment by nextgenusfs** [comment]
Okay, so phobius is failing. And to confirm this is local phobius install?

What happens when you run phobius from command line on one of the fasta files, i.e.:
```
phobius.pl -short KKMAI_000001.fa
```

**Comment by nextgenusfs** [comment]
Maybe the output is different in local versus remote (I don't have local phobius.pl on my current machine as I'm on a mac), can you post the output from `phobius.pl -short KKMAI_000001.fa`?  Or perhaps output from a few proteins would help as well.

**Comment by simas232** [comment]
I ran the local phobius installation for my experimental 100 proteins. An output file is attached

**Comment by nextgenusfs** [comment]
thanks.  Just to confirm, if you run with a single protein you would get the following results:
```
SEQENCE ID                     TM SP PREDICTION
KKMAI_000001                    0  0 o
```
```
SEQENCE ID                     TM SP PREDICTION
KKMAI_000006                    0  Y n4-15c20/21o
```

**Comment by simas232** [comment]
Yes, I tried to run phobius.pl for these two proteins and got exactly the same results.

**Comment by nextgenusfs** [comment]
Okay, thanks.  Hopefully try to get this figured out today.

**Comment by nextgenusfs** [comment]
Its plausible it is the python multiprocessing module.  Can you try this for me, move this script into /path/to/funannotate/util/phobius-multiproc-test.py.  And then run:

/path/to/funannotate/util/phobius-multiproc-test.py -i your_100_proteins.fa -o test_phobius.txt —debug

See if that runs correctly.

**Comment by nextgenusfs** [comment]
If that doesn't work, can you try this one and run in the same fashion:
https://www.dropbox.com/s/90hp2g5rqulnxcz/phobius-multiproc-test2.py?dl=0

I have to step out for a few hours, but will return to this later today.

**Comment by simas232** [comment]
I tried both scripts. Unfortunately, both of them didn't complete. That is, the folder with individual protein fasta files was generated, but results files *.phobius were not created. The log file phobius-multiproc.log also contained only the initial command, but not phobius.pl commands for individual proteins

**Comment by nextgenusfs** [comment]
Ok, let me make a few changes to get rid of trying to run multiple processes and see if that will work.

**Comment by nextgenusfs** [comment]
Okay, how about this one? This one should just run a single loop through the data and output a single output file.  https://www.dropbox.com/s/15qgm1m87kmfud2/phobius-multiproc-test3.py?dl=0

**Comment by simas232** [comment]
This script is also not finishing. Now I didn't get get the folder with individual protein fasta files. Just got the same log file phobius-multiproc.log, containing the first line "2017-06-10 18:02:50,607: /home/linuxbrew/funannotate/util/phobius-multiproc-test3.py -i kluyveromyces_marxianus_i.proteins.fa -o test_phobius.txt --debug" and two empty lines.

**Comment by nextgenusfs** [comment]
ok, we are nearly there.  yes you shouldn't have gotten the temp directory.  Can you confirm that this command runs to completion:
`phobius.pl -short kluyveromyces_marxianus_i.proteins.fa`

**Comment by simas232** [comment]
Yes, I ran this command again and got the same results as in phobius.txt, attached above.

**Comment by nextgenusfs** [comment]
Ok, I think i made a typo mistake in previous one.  fingers-crossed this one should do it.
https://www.dropbox.com/s/xom85mbm3q5jz02/phobius-multiproc-test4.py?dl=0

**Comment by simas232** [comment]
I got the same results as in test3. The log file phobius-multiproc.log also looks the same.

**Comment by nextgenusfs** [comment]
hmmm, well I will need to get on a box that has a local phobius install to see, not sure what else is wrong.  this one should print the output of phobius to terminal if output is empty, lets see if that works and/or if it prints any errors.  Basically all this script is doing right now is to launch `phobius.pl -short yourfasta.fa` from the python script, it should then be outputting a basename.phobius output file.

https://www.dropbox.com/s/o2ewav944t6t0iz/phobius-multiproc-test5.py?dl=0

**Comment by simas232** [comment]
I see. The results from test5 were the same as in test3 and test4. Here are the access permissions for of the files, just for comprehension:

-rwxr-xr-x 1 linuxbrew linuxbrew   3624 Jun 10 15:59 phobius-multiproc-test.py
-rwxr-xr-x 1 linuxbrew linuxbrew   3711 Jun 10 16:00 phobius-multiproc-test2.py
-rwxr-xr-x 1 linuxbrew linuxbrew   5120 Jun 10 18:01 phobius-multiproc-test3.py
-rwxr-xr-x 1 linuxbrew linuxbrew   5203 Jun 10 18:28 phobius-multiproc-test4.py
-rwxr-xr-x 1 linuxbrew linuxbrew   5369 Jun 10 18:46 phobius-multiproc-test5.py
-rwxr-xr-x 1 linuxbrew linuxbrew   3626 May 25 20:58 phobius-multiproc.py
-rwxr-xr-x 1 linuxbrew linuxbrew  565372 Jun 10 15:57 decodeanhmm
-rwxr-xr-x 1 linuxbrew linuxbrew   11397 Feb 14  2005 phobius.pl

**Comment by nextgenusfs** [comment]
Sorry if it seems like I'm asking a bunch of the same questions but you have phobius.pl in the $PATH correct?  You can run it from any folder and it works?

**Comment by nextgenusfs** [comment]
To expand on that, the output of 'which phobius.pl' is the version that you've modified to have the correct decodeanhmm?

**Comment by simas232** [comment]
Yes, I have a Docker image with funannotate. Phobius installation is included in PATH, it looks like this:

`/home/linuxbrew/phobius:/home/linuxbrew/.linuxbrew/opt/braker/libexec:/home/linuxbrew/funannotate:/home/linuxbrew/conda/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/linuxbrew/augustus/bin:/home/linuxbrew/.linuxbrew/opt/repeatmasker/libexec/util:/home/linuxbrew/gm_et_linux_64/gmes_petap:/home/linuxbrew/signalp-4.1:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`

When I try typing "which phobius.pl", I get response:

`/home/linuxbrew/phobius/phobius.pl`

which is indeed the installation place where I have _decodeanhmm_ modified.

**Comment by nextgenusfs** [comment]
Okay, finally getting somewhere.  actually has nothing to do with the script at all.  I spun up a docker container, and the remote phobius search works out of the box if you use `--email` option, it will fail if email is not given.  I think installed phobius locally, which was not easy as you mentioned the packaged decodeanhmm does not work.  So I then got decodeanhmm from tmhmm-2.0, from which only the decodeanhhmm.Linux_i386 worked for me.  The real problem here was actually how I was checking for the presence of `phobius.pl`, if you run from the command line `phobius.pl` it runs an infinite loop and the script stays open, so when the script was testing to see if local phobius was installed, it just got stuck and didn't go forward. 

If you replace the attached `library.py` file with the one in `/path/to/funannotate/lib/library.py` then it should probably work right away I'm hoping.  This fix will go into next release of funannotate. 

https://www.dropbox.com/s/gzx6h1mo4xhk7i9/library.py?dl=0

But I guess the real question is why didn't the remote phobius search work in the first place?  I'm not sure about that, other than the email argument is required I don't know why it would have failed, unless phobius.pl was in the $PATH and thus causing the infinite loop problem.

**Comment by simas232** [comment]
Yes, for me it also remains mysterious why remote phobius search doesn't work, although it used to work several months before. Firstly I thought that this may be due to the fact that I didn't have the local SignalP installation before, but removing it didn't solve the problem.

That's also interesting that only _decodeanhmm.Linux_i386_ worked for you. In my case, all four '86 versions worked fine, so I decided to use the newest version of it, since such processor architecture was released several years before 2004, when phobius was released.

Regarding the new version of library.py, I will do the clean run of _funannotate predict_ and _funannotate annotate_ and will let know about the results shortly.

**Comment by simas232** [comment]
The recent change of library.py solved all the problems! Now _funannotate annotate_ works perfectly.

Before running funannotate with the new version of library.py, I created the new funannotate Docker image from scratch. I fixed the local phobius installation with TMHMM's _decodeanhmm.Linux_i386_. There was no necessity to change the original phobius executable files access permissions.

Much thanks for the fix!

**Comment by nextgenusfs** [comment]
Good, glad it worked.  I should have thought of that little fix earlier.  I'm not a fan of docker and I don't think it works very well for funannotate, to build that docker image was a pain.  I think it is fine for testing purposes, but there are too many problems and it tends to be quite a bit slower than if you just install natively (of course depending on the OS you are on, etc).

**Comment by simas232** [comment]
Could be. I would also use use funannotate in my Macbook (OS X Sierra), but Perl module threads::shared cannot be installed normally due to the testing errors. I read that it's not good idea to proceed with the Perl module installations if tests are failing, so I didn't proceed to with installation. According to the information I was able to get in the Net, this is a bug is likely the bug from the developers of threads::shared package.

**Comment by nextgenusfs** [comment]
My "development" machine is MacBook pro and my bigger box is a Mac Pro with 24 cores.  So funannotate is written on a mac, support on linux is typically done through virtual box and/or docker.  I use home-brew perl and then cpanminus to install all the perl modules needed and haven't had a problem installing any of the modules.  More recently I've switched to using miniconda for python, as installation of the python modules is much easier on conda than with pip.

**Comment by nextgenusfs** [comment]
fixed in v0.7.0
